Initial work on integrating PONCHO into taskvine

BarrySlyDelgado opened the pull request:
For #2988, Notes:

- Added new flags to `poncho_package_run` to facilitate integration
1. `-n | --no-command`, This tells PONCHO not to run a command passed into the argument, passing in no command as an argument also does this. Thus, `poncho_package_run -n -e e.tar.gz python test.py` and `poncho_package_run -e e.tar.gz` will do the same thing.
2. `-k | --keep-alive`, This tells PONCHO to not activate the cleanup function on exit keeping the environment in place for the worker to use
3.  `-r | --remove-locking`, This tells PONCHO to unpack the environment without the use of file locking.

- One current issue is that activating the environment within poncho_package_run does not propagate up to the parent environment. It would probably be better to have it be done within `poncho_package_run`
- However, with `--keep-alive` the environment is still in place unpacked for the worker to use. Thus, added to `vine_cache.c` is the ability for the worker to activate the environment itself.
- Since we're using `keep-alive` PONCHO does not cleanup on exit, thus we might need the worker to do so.

dthain commented:
Does cleanup just involve deleting the directory tree, or is there more to it than that?

(The worker is responsible for deleting things no longer needed.)

btovar commented:
@BarrySlyDelgado When using -u DIR, the directory is not removed. I'd rather use -u DIR than add the --keep-alive facility, as now we have to figure out the tmp directory created. I think that the worker has to decide that DIR, it shouldn't be up to poncho.

btovar commented:
Also, I don't think there is a use for turning off locking. Either two instances are writing to two different directories, in which case locking has no effect, or they are writing to the same dir, in which case we are going to have hard to debug race conditions if locking is off.

btovar commented:
I'd suggest not to add the `-n | --no-command`. If -u is given, but no command is given, then we simply unpack. If -u or a command are missing, then we print an error message and exit.

dthain commented on taskvine/src/manager/taskvine.h:
```diff
@@ -44,6 +44,7 @@ typedef enum {
 	VINE_NOCACHE  = 0, /**< Do not cache file at execution site. (default) */
 	VINE_CACHE    = 1, /**< Cache file at execution site for later use. */
 	VINE_UNPACK   = 2, /**< Unpack this archive (.tar .tgz .zip) into a directory on arrival. */
+	VINE_PONCHO_UNPACK, /**< Unpacks a Poncho environment into a directory on arrival. */
```
Note that this enumeration is a bitfield where multiple flags can be set at once,
so each line needs a distinct value which is a power of two.

dthain commented on taskvine/src/worker/vine_cache.c:
```diff
@@ -228,6 +228,32 @@ int unpack_or_rename_target( struct cache_file *f, const char *transfer_path, co
 		unix_result = system(command);
 		free(command);
+	} else if(f-> flags & VINE_PONHO_UNPACK){
+		command = string_format("poncho_package_run -k -r -e %s --unpack-to %s", trasnfer_path, cache_path);
```
Don't forget to free the command.

dthain replied:
Actually, you could just make this a single command line string:
`poncho_package_run ... && activate && set_env`
since it all needs to run to completion anyway..

dthain commented on taskvine/src/worker/vine_cache.c:
```diff
@@ -228,6 +228,32 @@ int unpack_or_rename_target( struct cache_file *f, const char *transfer_path, co
+			debug(D_VINE,"command failed: %s",strerror(errno));
+			return 0;
+		}
+		command = string_format("%s/bin/activate", transfer_path);
```
Likewise

BarrySlyDelgado commented:
'set_env' is not present in all environments so the commands in `vine_cache.c`  were split into two to check if `set_env` exists.

- removed `no-command` and `remove-locking`.
- Added` free(command)` to `VINE_PONCHO_UNPACK` block
- Fixed bitfield enumeration
